Add `completions.lsp_insert_mode` setting to control what ranges are replaced when a completion is inserted

frederik-uni opened the pull request:
This PR adds `completions.lsp_insert_mode` and effectively changes the default from `"replace"` to `"replace_suffix"`, which automatically detects whether to use the LSP `replace` range instead of `insert` range.

`"replace_suffix"` was chosen as a default because it's more conservative than `"replace_subsequence"`, considering that deleting text is usually faster and less disruptive than having to rewrite a long replaced word.

Fixes #27197
Fixes #23395 (again)
Fixes #4816 (again)

Release Notes:

- Added new setting `completions.lsp_insert_mode` that changes what will be replaced when an LSP completion is accepted. The default is `"replace_suffix"`, but it accepts 4 values: `"insert"` for replacing only the text before the cursor, `"replace"` for replacing the whole text, `"replace_suffix"` that acts like `"replace"` when the text after the cursor is a suffix of the completion, and `"replace_subsequence"` that acts like `"replace"` when the text around your cursor is a subsequence of the completion (similiar to a fuzzy match). Check [the documentation](https://zed.dev/docs/configuring-zed#LSP-Insert-Mode) for more information.

frederik-uni committed:
somewhat smart mode

frederik-uni committed:
remove clone

marcospb19 reviewed (changes requested):
While I also prefer this behavior, we need to:

1. add a setting (`completions_mode`? `insertion_mode`?) with the variants `insertion`, `replace` and `smart` (or `auto`)
2. make `insertion` the default (unintended side effect: you won't have to change the current failing tests)
3. experiment for some time with `smart` before suggesting it as the new default

frederik-uni commented:
@marcospb19 
- I have no clue how to read the settings
- the default before was replace so any failing tests will still fail

SomeoneToIgnore commented:
Passing by and noting:

* this is quite a fundamental change, I'm not very comfortable seeing all these perturbations and 0 unit tests along the way
This includes https://github.com/zed-industries/zed/pull/27396 and the current PR.

* for a reference, I have thought that VSCode has `tab` and `enter` bound to `replace` and `insertion` modes correspondingly — that felt always enough to me and quite convenient as both options are available without tinkering with the settings. 
Should we mimic the same, ditch the `smart` mode and avoid altering settings, adding another action instead?

frederik-uni commented:
@marcospb19 suggested this feature. Personally I dont really care as I never really use the replace feature. That does probably apply to most other users as well. I dont see an issue with it tho.

frederik-uni committed:
add config

frederik-uni commented:
@marcospb19 to late

marcospb19 committed:
Test `CompletionMode::Insert` and `Replace`

marcospb19 committed:
Revert default `completion_mode` back to `replace`

marcospb19 committed:
update docs for `completion_mode`

marcospb19 committed:
rename `Smart` to `Auto`

marcospb19 committed:
Add tests for CompletionMode::Auto

marcospb19 committed:
style: rename variables to be more descriptive

marcospb19 committed:
Add CompletionMode::AutoSimilar

another mode that automatically decides whether to use `insert`
or `replace`, but, checks if suffix is a subsequence (fuzzy-ish) of the completion text, and not a strict suffix match

marcospb19 committed:
update docs

marcospb19 committed:
Make completion mode ReplaceSubsequence more conservative

Rename variants

Tweak tests

Co-authored-by: Max Brunsfeld <maxbrunsfeld@gmail.com>

marcospb19 committed:
update default to "replace_suffix"

Co-authored-by: Max Brunsfeld <maxbrunsfeld@gmail.com>

marcospb19 reviewed (approved):
I solved the conflicts and did more changes on top of your PR.

Added a bunch of tests.

And after talking with @maxbrunsfeld and @SomeoneToIgnore decided to change the default to the new `"replace_suffix"` variant, at least for now.

Thanks for the help!

marcospb19 committed:
clippy

marcospb19 committed:
more renaming

marcospb19 committed:
add website docs
